lib/charms/tls_certificates_interface/v3/tls_certificates.py

```diff
@@ -453,6 +455,8 @@ class ProviderCertificate:
     ca: str
     chain: List[str]
     revoked: bool
+    expiry_time: Optional[datetime] = None
```

Comment:
Is it possible for a certificate not to have an expiry time? If not, we shouldn't have this be an optional field here.

Reply:
Would it? Users don't instantiate this object, they only receive it. As far as I know this wouldn't break anything, but it's worth validating

Reply:
If we accept some assumptions like this one, then yes this shouldn't break anything, however the risk is still there, for example instantiating it in tests for mocking purposes.
I have changed it and I'm fine if we decide to merge it accepting the risk.

Reply:
I found out that Python dataclasses are not considered serialisable, and should not be used with that expectation.
serialising this object using json.dumps as an example would have succeeded until this point because all of its variables were serialisable so far, now with the change of adding a non serialisable member variable like datetime, any user serialising the object would now fail.
I would not consider this to be a breaking change since it was never promised that this object would be serialisable.
However I know for an example that the `self-signed-certificates-operator` would fail as it is doing json.dump of the object [here](https://github.com/canonical/self-signed-certificates-operator/blob/main/src/charm.py#L115).

@ghislainbourgeois, @gruyaume 
What do you think? It isn't difficult to solve the issue on the side of the implementing charm, but it won't be as smooth as fetching the new patch of the lib.

Reply:
I think we could add a `toJSON()` to the dataclass to make it simpler to serialize, I do not see any nicer way of solving the issue. It might be a breaking change in some cases, but as you said it was not something that was explicitly supported.

I think going with this as is, or adding a helper method would both be fine ways to handle this.